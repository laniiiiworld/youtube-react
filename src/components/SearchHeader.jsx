import React, { useEffect, useState } from 'react';
import { Link, useNavigate, useParams } from 'react-router-dom';
import { BsSunFill, BsMoonFill } from 'react-icons/bs';
import { FaYoutube, FaGithub } from 'react-icons/fa';
import { GoSearch } from 'react-icons/go';
import { useDarkMode } from '../context/DarkModeContext';

export default function SearchHeader() {
  const { darkMode, toggleDarkMode } = useDarkMode();
  const { keyword } = useParams();
  const [text, setText] = useState('');
  const navigate = useNavigate();

  const handleSubmit = (e) => {
    e.preventDefault();
    if (!text.trim()) return;
    navigate(`/videos/${text.trim()}`);
  };

  useEffect(() => setText(keyword || ''), [keyword]);

  return (
    <header className='w-full flex p-4 text-2xl border-b dark:border-zinc-600 mb-4 justify-between'>
      <Link to='/' className='flex items-center select-none'>
        <FaYoutube className='text-4xl text-brand' title='logo' />
        <h1 className='font-semibold ml-2 text-3xl tracking-tighter'>YouTube</h1>
      </Link>
      <form onSubmit={handleSubmit} className='w-5/12 flex items-center'>
        <input //
          type='text'
          placeholder='Search...'
          value={text}
          onChange={(e) => setText(e.target.value)}
          className='w-full p-2 pl-5 outline-none border-zinc-300 dark:bg-black dark:text-gray-500 border dark:border-zinc-900 rounded-s-full'
        />
        <button
          className='bg-zinc-100 hover:bg-zinc-200 border-zinc-300 dark:bg-zinc-600 hover:dark:bg-zinc-500 dark:border-zinc-900 px-5 py-3 border border-l-0 rounded-e-full'
          title='search'
        >
          <GoSearch />
        </button>
      </form>
      <div className='flex items-center gap-4'>
        <button
          onClick={() => toggleDarkMode()}
          title='dark/light mode'
          className={`text-2xl rounded-full p-1.5 ${
            darkMode ? '' : 'text-white bg-zinc-950 hover:bg-zinc-700'
          } dark:text-zinc-950 dark:bg-white dark:text-zinc-950 hover:dark:text-zinc-700`}
        >
          {darkMode ? <BsSunFill /> : <BsMoonFill />}
        </button>
        <button
          onClick={() => window.open(process.env.REACT_APP_GITHUB_URL)}
          title='github'
          className={`text-4xl rounded-full ${darkMode ? '' : 'hover:text-zinc-700'} hover:dark:bg-zinc-700`}
        >
          <FaGithub />
        </button>
      </div>
    </header>
  );
}
