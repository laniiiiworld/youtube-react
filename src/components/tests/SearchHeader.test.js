import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Route } from 'react-router-dom';
import { withAllContexts, withRouter } from '../../tests/utils';
import SearchHeader from '../SearchHeader';
import userEvent from '@testing-library/user-event';

describe('SearchHeader', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      withAllContexts(
        withRouter(<Route path='/' element={<SearchHeader />} />) //
      )
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with keyword correctly', () => {
    render(
      withAllContexts(
        withRouter(<Route path='/:keyword' element={<SearchHeader />} />, '/bts') //
      )
    );

    expect(screen.getByDisplayValue('bts')).toBeInTheDocument();
  });

  it('navigates to result page on search button clicked', () => {
    const keyword = 'fake keyword';
    render(
      withAllContexts(
        withRouter(
          <>
            <Route path='/' element={<SearchHeader />} />
            <Route path={`/videos/${keyword}`} element={<p>{`Search result for ${keyword}`}</p>} />
          </>
        )
      )
    );

    const searchInput = screen.getByPlaceholderText('Search...');
    const searchButton = screen.getByTitle('search');
    userEvent.type(searchInput, keyword);
    userEvent.click(searchButton);

    expect(screen.getByText(`Search result for ${keyword}`)).toBeInTheDocument();
  });
});
