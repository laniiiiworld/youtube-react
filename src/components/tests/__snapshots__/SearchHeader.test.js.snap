// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchHeader renders correctly 1`] = `
<header
  className="w-full flex p-4 text-2xl border-b dark:border-zinc-600 mb-4 justify-between"
>
  <a
    className="flex items-center select-none"
    href="/"
    onClick={[Function]}
  >
    <svg
      className="text-4xl text-brand"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 576 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        logo
      </title>
      <path
        d="M549.655 124.083c-6.281-23.65-24.787-42.276-48.284-48.597C458.781 64 288 64 288 64S117.22 64 74.629 75.486c-23.497 6.322-42.003 24.947-48.284 48.597-11.412 42.867-11.412 132.305-11.412 132.305s0 89.438 11.412 132.305c6.281 23.65 24.787 41.5 48.284 47.821C117.22 448 288 448 288 448s170.78 0 213.371-11.486c23.497-6.321 42.003-24.171 48.284-47.821 11.412-42.867 11.412-132.305 11.412-132.305s0-89.438-11.412-132.305zm-317.51 213.508V175.185l142.739 81.205-142.739 81.201z"
      />
    </svg>
    <h1
      className="font-semibold ml-2 text-3xl tracking-tighter"
    >
      YouTube
    </h1>
  </a>
  <form
    className="w-5/12 flex items-center"
    onSubmit={[Function]}
  >
    <input
      className="w-full p-2 pl-5 outline-none border-zinc-300 dark:bg-black dark:text-gray-500 border dark:border-zinc-900 rounded-s-full"
      onChange={[Function]}
      placeholder="Search..."
      type="text"
      value=""
    />
    <button
      className="bg-zinc-100 hover:bg-zinc-200 border-zinc-300 dark:bg-zinc-600 hover:dark:bg-zinc-500 dark:border-zinc-900 px-5 py-3 border border-l-0 rounded-e-full"
      title="search"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.25 2a8.25 8.25 0 0 1 6.34 13.53l5.69 5.69a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215l-5.69-5.69A8.25 8.25 0 1 1 10.25 2ZM3.5 10.25a6.75 6.75 0 1 0 13.5 0 6.75 6.75 0 0 0-13.5 0Z"
        />
      </svg>
    </button>
  </form>
  <div
    className="flex items-center gap-4"
  >
    <button
      className="text-2xl rounded-full p-1.5 text-white bg-zinc-950 hover:bg-zinc-700 dark:text-zinc-950 dark:bg-white dark:text-zinc-950 hover:dark:text-zinc-700"
      onClick={[Function]}
      title="dark/light mode"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"
        />
      </svg>
    </button>
    <button
      className="text-4xl rounded-full hover:text-zinc-700 hover:dark:bg-zinc-700"
      onClick={[Function]}
      title="github"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 496 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
        />
      </svg>
    </button>
  </div>
</header>
`;
